external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java

```diff
@@ -303,4 +315,38 @@ public void testLatestStrategyWhenTopologyIsRedeployed() {
         verify(collectorMock, never()).emit(anyList());
     }
 
+    @Test
+    public void testTimeStampStrategyWhenTopologyIsRedeployed() {
+        /**
+         * TIMESTAMP strategy should be applied if the emitter is new and the topology has been redeployed (storm id has changed)
+         * Offset should be reset according to the offset corresponding to startTimeStamp
+         */
+        long preRestartEmittedOffset = 20;
+        int preRestartEmittedRecords = 10;
+        long timeStampStartOffset = 2L;
+        long pollTimeout = 1L;
+        KafkaTridentSpoutBatchMetadata preExecutorRestartLastMeta = new KafkaTridentSpoutBatchMetadata(preRestartEmittedOffset, preRestartEmittedOffset + preRestartEmittedRecords - 1, "Some older topology");
+
+        Consumer kafkaConsumer = Mockito.mock(Consumer.class);
```

Comment:
This is a nitpick, but consider using partial mocking for this instead (https://static.javadoc.io/org.mockito/mockito-core/2.27.0/org/mockito/Mockito.html#partial_mocks), and wrap a real KafkaConsumer. That way you won't have to specify as much behavior for the consumer, and the test will be easier to maintain.

Reply:
I would have to replicate the exact same amount with the partial mock right 
1. add assignment or mock assignment
2. offsetsForTimes would have to be mocked anyway
3. add records or mock poll

Reply:
Yes, but it would be more robust to change. If there's an unexpected call to the stub (e.g. let's say we start calling `consumer.beginningOffsets` for some reason), the test will start failing. With a partial mock, we don't have to update the stub in that type of case.

The test is fine the way it is, we can replace it later if it becomes an issue.

Reply:
Aha. thanks for the explanation. That does make sense